### ISSUE TITLE
Tables with colspans generate incorrect DocBook XML

### ISSUE THREAD

**Issue by lnewson** [asciidoctor/asciidoctor#1462]
Given the following AsciiDoc syntax:
~~~
.Core Example
|====================================================
|               2+| Node 0        2+| Node 1
| Host processes  | Core 0 | Core 1 | Core 4 | Core 5
| Guest processes | Core 2 | Core 3 | Core 6 | Core 7
|====================================================
~~~
the output for DocBook 5.0 with AsciiDoctor 1.5.2 is (note the incorrect/duplicate colnames in the colspec and incorrect namest/nameend values):
~~~
<table frame="all" rowsep="1" colsep="1">
<title>Core Example</title>
<tgroup cols="5">
<colspec colname="col_1" colwidth="20*"/>
<colspec colname="col_2" colwidth="20*"/>
<colspec colname="col_2" colwidth="20*"/>
<colspec colname="col_3" colwidth="20*"/>
<colspec colname="col_3" colwidth="20*"/>
<tbody>
<row>
<entry align="left" valign="top"></entry>
<entry align="left" valign="top" namest="col_2" nameend="col_3"><simpara>Node 0</simpara></entry>
<entry align="left" valign="top" namest="col_3" nameend="col_4"><simpara>Node 1</simpara></entry>
</row>
<row>
<entry align="left" valign="top"><simpara>Host processes</simpara></entry>
<entry align="left" valign="top"><simpara>Core 0</simpara></entry>
<entry align="left" valign="top"><simpara>Core 1</simpara></entry>
<entry align="left" valign="top"><simpara>Core 4</simpara></entry>
<entry align="left" valign="top"><simpara>Core 5</simpara></entry>
</row>
<row>
<entry align="left" valign="top"><simpara>Guest processes</simpara></entry>
<entry align="left" valign="top"><simpara>Core 2</simpara></entry>
<entry align="left" valign="top"><simpara>Core 3</simpara></entry>
<entry align="left" valign="top"><simpara>Core 6</simpara></entry>
<entry align="left" valign="top"><simpara>Core 7</simpara></entry>
</row>
</tbody>
</tgroup>
</table>
~~~

From what I can see, the cause of this is that the column index is calculated incorrectly in lib/asciidoctor/table.rb as it assumes there are no previous colspans in the row. Below is a patch I believe will fix it by making use of the col_visits variable, however I'm not too familiar with ruby or the internals of AsciiDoctor so this may break something else (hence no pull request):
~~~
diff --git a/lib/asciidoctor/table.rb b/lib/asciidoctor/table.rb
index 587fc27..1130790 100644
--- a/lib/asciidoctor/table.rb
+++ b/lib/asciidoctor/table.rb
@@ -465,10 +465,10 @@ class Table::ParserContext
     1.upto(repeat) do |i|
       # make column resolving an operation
       if @col_count == -1
-        @table.columns << (column = Table::Column.new(@table, @current_row.size + i - 1))
+        @table.columns << (column = Table::Column.new(@table, @col_visits))
         if cell_spec && (cell_spec.has_key? 'colspan') && (extra_cols = cell_spec['colspan'].to_i - 1) > 0
           extra_cols.times do |j|
-            @table.columns << Table::Column.new(@table, @current_row.size + i + j - 1)
+            @table.columns << Table::Column.new(@table, @col_visits + j + 1)
           end
         end
       else
~~~

**Comment by mcrotty** [comment]
There are similar problems with a rowspan and attributes, eg. emphasis, applied to a column. 
Maybe the @col_visits needs to be call to effective_col_visits() so as the active rowspans are also taken into account.

**Comment by lnewson** [comment]
Tried on 1.5.3.dev and it's not working as mentioned above, the colspans are still generated incorrectly for the DocBook backend, as mentioned in the initial report:

~~~~
<table frame="all" rowsep="1" colsep="1">
<title>Core Example</title>
<tgroup cols="5">
<colspec colname="col_1" colwidth="20*"/>
<colspec colname="col_2" colwidth="20*"/>
<colspec colname="col_2" colwidth="20*"/>
<colspec colname="col_3" colwidth="20*"/>
<colspec colname="col_3" colwidth="20*"/>
<tbody>
<row>
<entry align="left" valign="top"></entry>
<entry align="left" valign="top" namest="col_2" nameend="col_3"><simpara>Node 0</simpara></entry>
<entry align="left" valign="top" namest="col_3" nameend="col_4"><simpara>Node 1</simpara></entry>
</row>
<row>
<entry align="left" valign="top"><simpara>Host processes</simpara></entry>
<entry align="left" valign="top"><simpara>Core 0</simpara></entry>
<entry align="left" valign="top"><simpara>Core 1</simpara></entry>
<entry align="left" valign="top"><simpara>Core 4</simpara></entry>
<entry align="left" valign="top"><simpara>Core 5</simpara></entry>
</row>
<row>
<entry align="left" valign="top"><simpara>Guest processes</simpara></entry>
<entry align="left" valign="top"><simpara>Core 2</simpara></entry>
<entry align="left" valign="top"><simpara>Core 3</simpara></entry>
<entry align="left" valign="top"><simpara>Core 6</simpara></entry>
<entry align="left" valign="top"><simpara>Core 7</simpara></entry>
</row>
</tbody>
</tgroup>
</table>
~~~~

Also I have no idea why @mcrotty is mentioning rowspans, as it's never mentioned in this report. I do agree though that you shouldn't be doing rowspans on the header.

**Comment by mcrotty** [comment]
A correction, *your patch* works on 1.5.3.dev, and the logic looks sound.
Rowspan comes into it where the column "index" is calculated ie. whether the index is @col_visits or effective_col_visits()). As there can be no rowspans in the "header" then @col_visits is okay. 
Hope that clarifies the comment above.

**Comment by lnewson** [comment]
Ahh yep, that makes a lot more sense and good catch on the rowspans. Thanks for the clarification @mcrotty :)

**Comment by lnewson** [comment]
Looked at this again since 1.5.3/1.5.4 has been released and it looks to have been resolved. Looking at the changes, it looks like the fix for #1556 is what fixed it (though to me that issue looks like a duplicate of this one). Anyways feel free to close this as it looks fixed from my tests.

**Comment by mojavelinux** [comment]
That's great news! Thanks for testing. I went ahead and add your example as a test just to be certain.
